File: 2104-子数组范围和.cpp
#include <iostream>
using namespace std;
#include <algorithm>
#include <vector>
#include <stack>

class Solution
{
public:
    long long subArrayRanges(vector<int> &nums)
    {
        // 暴力
        // long long res = 0;
        // for (int i = 0; i < nums.size(); i++)
        // {
        //     int min_v = INT_MAX;
        //     int max_v = INT_MIN;
        //     for (int j = i; j < nums.size(); j++)
        //     {
        //         min_v = min(min_v, nums[j]);
        //         max_v = max(max_v, nums[j]);
        //         res += max_v - min_v;
        //     }
        // }
        // return res;
        // 单调栈
        int n = nums.size();
        vector<int> min_l(n), min_r(n), max_l(n), max_r(n);
        stack<int> min_stack, max_stack;
        for (int i = 0; i < n; i++)
        {
            int num = nums[i];
            while (!min_stack.empty() && num < nums[min_stack.top()])
            {
                min_stack.pop();
            }
            min_l[i] = min_stack.empty() ? -1 : min_stack.top();
            min_stack.push(i);
            while (!max_stack.empty() && num >= nums[max_stack.top()])
            {
                max_stack.pop();
            }
            max_l[i] = max_stack.empty() ? -1 : max_stack.top();
            max_stack.push(i);
        }
        min_stack = stack<int>();
        max_stack = stack<int>();
        for (int i = n - 1; i >= 0; i--)
        {
            int num = nums[i];
            while (!min_stack.empty() && num <= nums[min_stack.top()])
            {
                min_stack.pop();
            }
            min_r[i] = min_stack.empty() ? n : min_stack.top();
            min_stack.push(i);
            while (!max_stack.empty() && num > nums[max_stack.top()])
            {
                max_stack.pop();
            }
            max_r[i] = max_stack.empty() ? n : max_stack.top();
            max_stack.push(i);
        }
        long long min_sum = 0, max_sum = 0;
        for (int i = 0; i < n; i++)
        {
            int num = nums[i];
            min_sum += (long long)(i - min_l[i]) * (min_r[i] - i) * num;
            max_sum += (long long)(i - max_l[i]) * (max_r[i] - i) * num;
        }
        return max_sum - min_sum;
    }
};

int main()
{
    vector<int> nums = {4, -2, -3, 4, 1};
    Solution sol;
    long long res = sol.subArrayRanges(nums);
    cout << res << endl;
}

File: 2104-子数组范围和.py
from typing import List


class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        # 暴力
        # res = 0
        # if len(nums) <= 1:
        #     return res
        # for i in range(len(nums)):
        #     min_v, max_v = float('inf'), -float('inf')
        #     for j in range(i, len(nums)):
        #         min_v = min(nums[j], min_v)
        #         max_v = max(nums[j], max_v)
        #         res += max_v - min_v
        # return res
        # 单调栈
        n = len(nums)

        min_l, max_l = [0]*n, [0]*n
        min_stack, max_stack = [], []
        for i, num in enumerate(nums):
            # 单调递增栈，得到左侧最近的比num小的数
            while min_stack and num < nums[min_stack[-1]]:
                min_stack.pop()
            min_l[i] = min_stack[-1] if min_stack else -1
            min_stack.append(i)
            # 单调递减栈，得到左侧最近的比num大的数
            while max_stack and num >= nums[max_stack[-1]]:
                max_stack.pop()
            max_l[i] = max_stack[-1] if max_stack else -1
            max_stack.append(i)

        min_r, max_r = [0]*n, [0]*n
        min_stack, max_stack = [], []
        for i in range(n-1, -1, -1):
            num = nums[i]
            # 单调递增栈，得到右侧最近的比num小的数
            while min_stack and num <= nums[min_stack[-1]]:
                min_stack.pop()
            min_r[i] = min_stack[-1] if min_stack else n
            min_stack.append(i)
            # 单调递减栈，得到右侧最近的比num大的数
            while max_stack and num > nums[max_stack[-1]]:
                max_stack.pop()
            max_r[i] = max_stack[-1] if max_stack else n
            max_stack.append(i)

        min_sum = max_sum = 0
        for i, num in enumerate(nums):
            min_sum += (i - min_l[i]) * (min_r[i] - i) * num
            max_sum += (i - max_l[i]) * (max_r[i] - i) * num
        return max_sum - min_sum


nums = [4, -2, -3, 4, 1]
sol = Solution().subArrayRanges(nums)
print(sol)
